Precondition: run Task2_clean.ipynb before running this code.
This creates the input file of the Customer-Product utility matrix

This code performs a collaborative based item recommendation system.
This approach is a memory based solution

In [9]:
import pandas as pd
import numpy as np
from product_data import ProductData

In [10]:
file_name = "recom_pivot.csv"
prod = ProductData(file_name)
prod.set_pivot_dataframe_data_types()
prod.df_orig_recommender.head()
prod.df_products.head()

5000.5  5001.0  5001.5  5002.0  5002.5  5003.0  5004.0  5004.5  5005.0  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   
2       0       0       0       0       0       0       0       0       0   
3       0       0       0       0       0       0       0       0       0   
4       0       0       0       1       0       0       0       0       0   

   5007.0  ...  200035.5  200037.0  200037.5  200038.0  200038.5  200045.5  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   200046.0  200046.5  200047.0  350027.5  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 333 columns]

In [11]:
def output_results(selected_product, df_corr):
    df_single_product = df_corr[selected_product].to_frame()
    df_sort = df_single_product.fillna(0).sort_values(by=[selected_product], ascending=False)
    print(f"If you have bought {selected_product} then you may want to consider the following products:")
    print(df_sort.head())
    print("--------------------------")




In [12]:
selected_products = ["200019.5", "5001.0", "5001.5", "5002.0", "5000.5"]
df_corr = prod.df_products.corr()
for selected_product in selected_products:
    output_results(selected_product, df_corr)

If you have bought 200019.5 then you may want to consider the following products:
          200019.5
200019.5  1.000000
200035.5  0.104986
200021.0  0.083080
200005.0  0.035654
35088.0   0.030168
--------------------------
If you have bought 5001.0 then you may want to consider the following products:
           5001.0
5001.0   1.000000
5000.5   0.153768
10023.5  0.147058
5002.0   0.140715
10006.0  0.139063
--------------------------
If you have bought 5001.5 then you may want to consider the following products:
           5001.5
5001.5   1.000000
40077.5  0.098042
5000.5   0.084167
5001.0   0.063253
35094.5  0.062200
--------------------------
If you have bought 5002.0 then you may want to consider the following products:
           5002.0
5002.0   1.000000
5001.0   0.140715
5000.5   0.136879
10032.5  0.128552
5011.5   0.125499
--------------------------
If you have bought 5000.5 then you may want to consider the following products:
           5000.5
5000.5   1.000000
25003.0  0.19705

If a person buys product 5001.0, then the recommender systems suggests that that person may want to also buy the products 5000.5, 10023.5, 5002.0, 10006.0. These products have a low selling correlation with product 5001.0 of less than 0.2 and may not be very good choices. The same applies (ie weak correlation) for the other products tested.
